scalyr_agent/copying_manager/worker.py

```diff
@@ -1021,6 +1036,17 @@ def __scan_for_new_bytes(self, current_time=None):
         for processor in self.__log_processors:
             processor.scan_for_new_bytes(current_time)
 
+    def get_full_checkpoints(self):  # type: () -> Dict[six.text_type, Any]
+        """
+        This method returns the full collection of the current log processor checkpoint states.
+        It also mainly designed to be called from the outside threads, so the checkpoint object should be protected
+        by the lock.
+        :return: the dict of the checkpoint states of the current log processors.
+        """
+        with self._checkpoints_lock:
+            # this variable is updated when the '__write_checkpoint_state' method is writing to the checkpoint file.
+            return self._full_checkpoints
```

Comment:
I think we also need to add a comment either in `__write_checkpoint_state` or where `_full_checkpoints` is declared that the you can only modify `_full_checkpoints` by replacing a reference to it.  I say that because we are just getting a reference to the dictionary while holding the lock.  Once we release the lock, someone could modify the dictionary pointed to by `_full_checkpoints` and screw us up.

The safer approach is to just return a clone of `_full_checkpoints` here.  Which, maybe we should do.  But, it's more copying.